### PR TITLE
[WIP] non-root user for NodeJS template

### PR DESCRIPTION
**Pull request by austinfrey**
Initial draft of non-root user in the NodeJS template.

## Description

update the template/node/Dockerfile to use the `node` user rather than root. function directory structure needed to changed from `/root/` to `/usr/local/app/` as well 

## Motivation and Context

Allows users to run function as non-root which may increase security

This relates to #81 
- [x] I have raised an issue to propose this change ([required](https://github.com/alexellis/faas/blob/master/CONTRIBUTING.md))
already raised by @alex #81 

## How Has This Been Tested?

Manually built test function after changes, deployed and tested through the UI gateway

Linux 16+ on Docker CE 17.06

Will only impact the NodeJS template

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/alexellis/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by aafrey** [comment]
Initial draft to get feedback before moving on to other templates. No need to merge yet.

File: template/node/Dockerfile [hunk 1]
```diff
@@ -12,27 +12,31 @@ WORKDIR /root/
 # Turn down the verbosity to default level.
 ENV NPM_CONFIG_LOGLEVEL warn
 
+RUN mkdir -p /usr/local/app/function
```
Comment (alexellis) [inline template/node/Dockerfile]: What about a regular home directory like `/home/app/` ?

File: template/node/Dockerfile [hunk 2]
```diff
@@ -12,27 +12,31 @@ WORKDIR /root/
...
-WORKDIR /root/
-COPY function           function
-WORKDIR /root/function
+WORKDIR /usr/local/app/
+COPY function           ./function
 
-WORKDIR /root/
+WORKDIR /usr/local/app/
+
+USER node
```
Comment (alexellis) [inline template/node/Dockerfile]: Great start! Maybe the username could be `app` and we could use that across the board - like a copy/paste segment for all the templates?

File: template/node/Dockerfile (hunk 1 above)
Comment (aafrey) [reply template/node/Dockerfile]: sounds good

File: template/node/Dockerfile (hunk 2 above)
Comment (aafrey) [reply template/node/Dockerfile]: that makes the most sense. I'll work on it this evening

**Comment by alexellis** [comment]
@aafrey - really appreciate your work on this and the "serverless inc" template. 💯 

When can we get this fixed-up and merged? @rgee0 @rorpage any other feedback?

File: template/node/Dockerfile (hunk 2 above)
Comment (rgee0) [reply template/node/Dockerfile]: Just picking up on the user, and acknowledging that its largely academic, but why dont we use `faas` or `openfaas`?  Setting this as a sort of principle might be helpful further down the road, when 3rd party templates start to emerge.

**Comment by aafrey** [comment]
I'll allocate some time tomorrow for this if no one else has any feedback @alex @rgee0 @rorpage

File: template/node/Dockerfile (hunk 2 above)
Comment (alexellis) [reply template/node/Dockerfile]: I'd just keep it generic like `app` for now. Always good to think forward though so thanks for your input.

**Comment by ems5311** [comment]
When I try this change, I get the following error:

```
$ curl http://localhost:8080/function/nodejs-echo -d "stuff"
exit status 1
fs.js:641
  return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);
                 ^

Error: EACCES: permission denied, open '/usr/local/app/index.js'
    at Error (native)
    at Object.fs.openSync (fs.js:641:18)
    at Object.fs.readFileSync (fs.js:509:33)
    at Object.Module._extensions..js (module.js:578:20)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:389:7)
    at startup (bootstrap_node.js:149:9)
```

I believe this can be fixed by the following line just before the `USER node` line:

`RUN chown -R node .`

**Comment by alexellis** [comment]
@ems5311 thank you for testing.

**Comment by aafrey** [comment]
@ems5311 appreciate the feedback. I couldn't reproduce so I'm unsure what the issue is. I wouldn't think you'd need to `chown` since the `node` user is setup by default when building from that image.  I made some changes as asked to use `/home/app` directory and the `app` user. I tested and it seems to be working for me, but it worked for me last time too, so would you mind testing again, and let me know if you see that error again? i thought docker was supposed to fix the "works on my machine" issue :)

**Comment by ems5311** [comment]
@aafrey Thanks for the feedback, strangely I am still seeing the issue. I have uploaded the image that `faas-cli build -f <yaml>` builds for me for the nodejs-echo function, based on this latest dockerfile. https://hub.docker.com/r/ems5311/nodejs-echo-noroot/

I believe if you run `faas-cli deploy --image ems5311/nodejs-echo-noroot --name ems5311-node-noroot`, it will pull and run the image from the docker hub repo.

I have also made sure that I am building the `faas-cli` binary with this latest pull request, and the `faas-cli version` hash matches the latest commit. I am running Ubuntu 16.04 and I might have some unconventional docker settings on my system.

Thanks again,
Eric

**Comment by aafrey** [comment]
@ems5311 Hi Eric, the image you sent over doesn't work for me either, could you try building with the `build.sh` script in the `template/node` directory rather than the cli? just to test if there is a difference

**Comment by ems5311** [comment]
@aafrey good call, I'll try this out tonight.

**Comment by ems5311** [comment]
Yep, works now. Force rebuilding the image with `--no-cache` did the trick. Sorry for the false alarm, this changelist is good to go. :clap:

**Comment by aafrey** [comment]
@ems5311 appreciate the testing! thanks!

File: template/node/Dockerfile (hunk 1 above)
Comment (alexellis) [reply template/node/Dockerfile]: This should be: `/home/app/` as per the comment, could you update? I'd like to move forward and merge your change. Thank you for helping lead this

**Comment by aafrey** [comment]
@alexellis anything else needed here?

**Comment by alexellis** [comment]
Yes I posted a comment about using /home/app as the folder for the app.

**Comment by aafrey** [comment]
the most recent commit made that change

**Comment by alexellis** [comment]
@austinfrey are we good to go with this?

File: template/node/Dockerfile [hunk 3]
```diff
@@ -12,27 +12,30 @@ WORKDIR /root/
...
+WORKDIR /home/app/function
 RUN npm i || :
-WORKDIR /root/
-COPY function           function
-WORKDIR /root/function
+WORKDIR /home/app/
+COPY function           ./function
 
-WORKDIR /root/
+RUN addgroup -S app && adduser -S -g app app
```
Comment (alexellis) [inline template/node/Dockerfile]: We can add the group at the top - meaning this gets cached and not affected by copying code in.

Comment (austinfrey) [reply template/node/Dockerfile]: i can make this change @alexellis then we should be all set. should I make the change, squash and resubmit? or can i squash and commit with this same PR?

**Comment by alexellis** [comment]
Yes you can squash and re-push with the --force flag. I only raise a new PR if I screw things up really bad and have to push out of a different branch.

**Comment by alexellis** [comment]
Ping me for merging.

**Comment by austinfrey** [comment]
@alexellis i had to merge the latest master with my branch to get up to date, i squashed everything down to one commit, but attributing all the previous changes i had to merge in as being part of this commit. is that normal?

**Comment by alexellis** [comment]
Struggling to see the changes here. This contains 32 files and now has some merge conflicts. Maybe it would be easier to submit a new PR?

**Comment by austinfrey** [comment]
Agreed, I'll resubmit this evening

**Comment by austinfrey** [comment]
@alexellis started fresh, hopefully that's the last of it :)

**Comment by alexellis** [comment]
Derek add label: status/testing

File: template/node/Dockerfile [hunk 4]
```diff
@@ -1,38 +1,41 @@
...
 
+RUN addgroup -S app && adduser -S -g app app
 WORKDIR /root/
 
 # Turn down the verbosity to default level.
 ENV NPM_CONFIG_LOGLEVEL warn
 
+RUN mkdir -p /home/app
+
 # Wrapper/boot-strapper
```
Comment (alexellis) [inline template/node/Dockerfile]: Just wondering if "user app" can go higher?

Comment (austinfrey) [reply template/node/Dockerfile]: sure. I'll update and test

**Comment by austinfrey** [comment]
@alexellis moved `adduser` to line two. tested and had no issues

**Comment by alexellis** [comment]
Merged. Please can you retro-fit to the armhf template? @rgee0 perhaps you could help test?

**Comment by alexellis** [comment]
```
faas-cli invoke notroot
Reading from STDIN - hit (Control + D) to stop.
test this
Server returned unexpected status code: 500 - exit status 1
fs.js:641
  return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);
                 ^

Error: EACCES: permission denied, open '/home/app/index.js'
    at Error (native)
    at Object.fs.openSync (fs.js:641:18)
    at Object.fs.readFileSync (fs.js:509:33)
    at Object.Module._extensions..js (module.js:578:20)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:389:7)
    at startup (bootstrap_node.js:149:9)
```

@austinfrey this broke :-/

**Comment by austinfrey** [comment]
Interesting. I'll look at this tonight

**Pull request merged**